src/cs/Ssh.Tcp/Services/StreamForwarder.cs

```diff
@@ -0,0 +1,214 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.DevTunnels.Ssh.Tcp;
+
+public class StreamForwarder : IDisposable
+{
+	private bool disposed;
+
+	public StreamForwarder(
+		Stream localStream,
+		Stream remoteStream,
+		TraceSource trace)
+	{
+		LocalStream = localStream ?? throw new ArgumentNullException(nameof(localStream));
+		RemoteStream = remoteStream ?? throw new ArgumentNullException(nameof(remoteStream));
+		Trace = trace ?? throw new ArgumentNullException(nameof(trace));
+
+		ForwardStreamAsync();
+	}
+
+	public Stream LocalStream { get; }
+
+	public Stream RemoteStream { get; }
+
+	public TraceSource Trace { get; }
+
+	public event EventHandler<EventArgs>? Closed;
+
+	private async void ForwardStreamAsync()
+	{
+		try
+		{
+			bool endOfStream = await ForwardStreamAsync(LocalStream, RemoteStream)
+				.ConfigureAwait(false);
+			Close(abort: !endOfStream);
+		}
+		catch (Exception ex)
+		{
+			// Catch all exceptions in this async void method.
+			Trace.TraceEvent(
+				TraceEventType.Error,
+				SshTraceEventIds.UnknownError,
+				$"{nameof(PortForwardingService)} unexpected error forwarding stream: {ex}");
+			Close(abort: true);
```

Comment:
If `Closed` event handler throws, it will kill the process. It's not caught and inside `async void`.

Reply:
Same for `Trace.TraceEvent()`. If trace throws (it can throw `ObjectDisposedException` for example), it'll also kill the process. Async voids are dangerous :)

Reply:
Updated.